### PR TITLE
tracing: trace sys_init calls

### PR DESCRIPTION
**Pull request by ycsin**
The init functions has `int` as return type, but currently there is no way to know the result of the `SYS_INIT`s.

This PR creates additional tracing APIs to trace the enter and (exit + result) of SYS_INIT and DEVICE_DT_DEFINE (and friends), which can be useful to debug things.

### CODE REVIEW

**Review by luchnikov — DISMISSED** [review]

**Review by jgl-meta — DISMISSED** [review]
lgtm

**Review by andyross — DISMISSED** [review]
Sure.  With the caveat that trace points are historically fragile and keep getting disabled or confuzzled as code moves around.

FWIW, since you've been picking up so much general slack: I continue to believe we want to scrap this trace framework in favor of something that works in an automated way, e.g. with a little linker support to add stubs to function entry, and/or preprocessor indirection to add stubs to the start and end of a function definition, etc...  Non-trivial but probably fun.

**Review by cfriedt — DISMISSED** [review]

**Comment by ycsin** [comment]
I've added the `level` arg to the exit trace function to make it more helpful, PTAL, thanks!

**Review by cfriedt — APPROVED** [review]

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

**Comment by ycsin** [comment]
> Exceptions require approval by the release team and a vote by the TSC.
> https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

I'd like to request for a review in the TSC if possible, the change in this PR is pretty straightforward and will enable LTS users to trace the init functions.

**Comment by ycsin** [comment]
> I continue to believe we want to scrap this trace framework in favor of something that works in an automated way, e.g. with a little linker support to add stubs to function entry, and/or preprocessor indirection to add stubs to the start and end of a function definition, etc...

Sounds interesting but I do not have much experience in that area, do you mean something like `--wrap` or `-finstrument-functions`?

**Comment by aescolar** [comment]
> I'd like to request for a review in the TSC if possible, the change in this PR is pretty straightforward and will enable LTS users to trace the init functions.

Please join the release meeting tomorrow, so we can talk about it
CC @nashif

**Review by nashif — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
I'd like to see this go into v3.7.0. While we still have the known issue of not being able to programmatically inspect the return value of `SYS_INIT()` functions, at least this allows tracing to reveal the return code.

**Pull request merged**
